<img src="images/logo.jpg" style="display: block; margin-left: auto; margin-right: auto;" alt="לוגו של מיזם לימוד הפייתון. נחש מצויר בצבעי צהוב וכחול, הנע בין האותיות של שם הקורס: לומדים פייתון. הסלוגן המופיע מעל לשם הקורס הוא מיזם חינמי ללימוד תכנות בעברית.">

# <p style="text-align:right;direction:rtl;"> תרגילים </p> 

## <p style="text-align:right;direction:rtl;"> בחירות במדינת הגמדים </p>

<p style="text-align:right;direction:rtl;">
    במדינת הגמדים רעש מהומה, ולכן החליטו ללכת לבחירות.<br>
    ועדת הבחירות המרכזית, בראשות דרדסבא, ערכה בחירות חשאיות והוגנות והציגה לפניכם את התוצאות בקובץ elections.txt המצורף (בתיקיית resources), על כל 353,400 בוחריה.<br>
    הקובץ הינו רשימה של שמות המפלגות שבהן בחרו תושבי מדינת הגמדים, כאשר כל שורה מייצגת בחירה של תושב מדינת הגמדים.<br>
    כדי לשמור על טוהר הבחירות לא נאמר לכם במפורש מהו מספר המפלגות שהגישו את רשימותיהן לוועדת הבחירות.<br>
    עליכם הוטלה המשימה לעבור על קובץ תוצאות הבחירות ולדווח כמה קולות קיבלה כל מפלגה.<br>
    כתבו פונקציה שלא מקבלת פרמטרים ומחזירה את תוצאות הבחירות במילון ובו שם המפלגה ומספר הקולות שקיבלה.
</p>

In [14]:
# כתבו את הפונקציה שלכם כאן: get_election_results()
def get_election_results():
    with open('resources\elections.txt', 'r') as file_elections:
        temp = file_elections.readlines()
    length_temp = len(temp)
    i = 0
    dictionariy ={}
    one = 1
    while length_temp:
        new = temp[i]
        new = new.strip()
        if new not in dictionariy:
            dictionariy.update({new:one})
        else:
            new_new = dictionariy.get(new)
            new_new += 1
            dictionariy.update({new:new_new})
        i += 1
        length_temp -= 1
    return dictionariy

get_election_results()

{'OnlyTovTov': 41075,
 'NilsIsALeader': 63702,
 'DorothyForMidgeLand': 23071,
 'MidgeLandIsUs': 68582,
 'SnufkinForCouncil': 35830,
 'WiliWanka': 26628,
 'GulliverIsBest': 25928,
 'BestForMidgeLand': 40144,
 'FingeroniBringsPeace': 22457,
 'GreenMidgeLand': 5983}

In [15]:
def get_election_results():
    with open('resources\elections.txt', 'r') as file_elections:
        temp = file_elections.readlines()
        file_elections.seek(0)
        temp_new = file_elections.read()
    length_temp = len(temp)
    i = 0
    dictionariy ={}
    one = 1
    while length_temp:
        new = temp[i]
        new = new.strip()
        if new not in dictionariy:
            count = temp_new.count(str(new))
            dictionariy.update({new:count})
        i += 1
        length_temp -= 1
    return(dictionariy)
get_election_results()

{'OnlyTovTov': 41075,
 'NilsIsALeader': 63702,
 'DorothyForMidgeLand': 23071,
 'MidgeLandIsUs': 68582,
 'SnufkinForCouncil': 35830,
 'WiliWanka': 26628,
 'GulliverIsBest': 25928,
 'BestForMidgeLand': 40144,
 'FingeroniBringsPeace': 22457,
 'GreenMidgeLand': 5983}

<p style="text-align:right;direction:rtl;">
    אחוז החסימה במדינת הגמדים הוא אכזרי למדי ועומד על 10%.<br>
    כיון שהצביעו 353,400 גמדים – אחוז החסימה עומד על
    $353,400*0.1=35,340$ קולות.<br>
    כתבו פונקציה המקבלת את מילון תוצאות הבחירות ומדפיסה אילו מפלגות עברו את אחוז החסימה ואילו לא.
</p>

In [23]:
PERCENT_BLOCKED = 353400 * 0.1
def pass_the_percent_blocked(dictionariy):
    counter = 0
    for key, val in dictionariy.items():
        if val > PERCENT_BLOCKED:
            print(val)
            print('pass: ' + key + 'with: ' + str(val))
            counter += val
        else:
            print('not pass: ' + key + 'with: '+ str(val))
    return counter


dictionariy = get_election_results()
pass_the_percent_blocked(dictionariy)

41075
pass: OnlyTovTovwith: 41075
63702
pass: NilsIsALeaderwith: 63702
not pass: DorothyForMidgeLandwith: 23071
68582
pass: MidgeLandIsUswith: 68582
35830
pass: SnufkinForCouncilwith: 35830
not pass: WiliWankawith: 26628
not pass: GulliverIsBestwith: 25928
40144
pass: BestForMidgeLandwith: 40144
not pass: FingeroniBringsPeacewith: 22457
not pass: GreenMidgeLandwith: 5983


249333

<p style="text-align:right;direction:rtl;">
כעת כתבו פונקציה המקבלת את מילון תוצאות הבחירות ומחזירה את מספר הקולות שקיבלו רק מפלגות שעברו את אחוז החסימה.</p>

In [26]:
PERCENT_BLOCKED = 353400 * 0.1


def get_election_results():
    with open('resources\elections.txt', 'r') as file_elections:
        temp = file_elections.readlines()
        file_elections.seek(0)
        temp_new = file_elections.read()
    length_temp = len(temp)
    i = 0
    dictionariy ={}
    one = 1
    while length_temp:
        new = temp[i]
        new = new.strip()
        if new not in dictionariy:
            count = temp_new.count(str(new))
            dictionariy.update({new:count})
        i += 1
        length_temp -= 1
    return(dictionariy)


def pass_the_percent_blocked(dictionariy):
    counter = 0
    for key, val in dictionariy.items():
        if val > PERCENT_BLOCKED:
            print('pass: ' + key + 'with: ' + str(val))
            counter += val
        else:
            print('not pass: ' + key + 'with: '+ str(val))
    return counter


def number_of_votes_are_pass(dictionariy):
    return (pass_the_percent_blocked(dictionariy))


dictionariy = get_election_results()
print(number_of_votes_are_pass(dictionariy))    

pass: OnlyTovTovwith: 41075
pass: NilsIsALeaderwith: 63702
not pass: DorothyForMidgeLandwith: 23071
pass: MidgeLandIsUswith: 68582
pass: SnufkinForCouncilwith: 35830
not pass: WiliWankawith: 26628
not pass: GulliverIsBestwith: 25928
pass: BestForMidgeLandwith: 40144
not pass: FingeroniBringsPeacewith: 22457
not pass: GreenMidgeLandwith: 5983
249333


<p style="text-align:right;direction:rtl;">
    אחרי שסיימתם בדקו שקיבלתם את הערך הבא: <span style="direction: rtl; background: #000; text: #000">249,333</span>
</p>

<p style="text-align:right;direction:rtl;">
    לאחר קבלת תוצאות האמת המפלגות פועלות להקים קואליציה. קואליציה מורכבת מ־61 מנדטים לפחות.<br>
    מנדט שווה: $\dfrac{votes}{120}=2.077.775\cong2,078$ קולות.<br>
    שימו לב, הקולות בתרגיל זה הם התוצאה של הסעיף הקודם. כלומר הקולות בניכוי הקולות עבור מפלגות שלא עברו את אחוז החסימה.<br>
    כעת כתבו פונקציה המקבלת את מילון התוצאות ומחזירה מילון שבו שם המפלגה ממופה עם מספר המנדטים שקיבלה המפלגה.<br>
    הפעילו את פונקנציית <code>round</code> על תוצאת החלוקה (השתמשו ב־2,078) כדי לקבל את מספר המנדטים השלם.
</p>

In [66]:
# Upload 67
PERCENT_BLOCKED = 353400 * 0.1
MANDAT = 2078


def get_election_results():
    with open('resources\elections.txt', 'r') as file_elections:
        temp = file_elections.readlines()
        file_elections.seek(0)
        temp_new = file_elections.read()
    length_temp = len(temp)
    i = 0
    dictionariy = {}
    while length_temp:
        new = temp[i]
        new = new.strip()
        if new not in dictionariy:
            count = temp_new.count(str(new))
            dictionariy.update({new: count})
        i += 1
        length_temp -= 1
    return(dictionariy)


def pass_the_percent_blocked(dictionariy):
    counter = 0
    for key, val in dictionariy.items():
        if val > PERCENT_BLOCKED:
            print('pass: ' + key + 'with: ' + str(val))
            counter += val
        else:
            print('not pass: ' + key + 'with: ' + str(val))
    return counter


def number_of_votes_are_pass(dictionariy):
    return (pass_the_percent_blocked(dictionariy))


def calc_mandatim(dictionariy):
    for key, val in dictionariy.items():
        if val > PERCENT_BLOCKED:
            val = round(val / MANDAT)
            print(key + ' ' + str(val))
            dictionariy.update({key: val})
        else:
            dictionariy.update({key: 0})
            # print(dictionariy.get(key))
    return dictionariy


def coalition_train(dictionariy):
    coalition_one = ''
    sum_one = 0
    coalition_two = ''
    sum_two = 0
    for key in dictionariy:
        if key == 'NilsIsALeader':
            for key_s, value in dictionariy.items():
                if (value != 0) and (key_s != 'MidgeLandIsUs') and (sum_one < 61):
                    coalition_one += key_s + ' '
                    sum_one += value
        if key == 'MidgeLandIsUs':
            for key_s, value in dictionariy.items():
                if (value != 0) and (key_s != 'NilsIsALeader') and (sum_two < 61):
                    coalition_two += key_s + ' '
                    sum_two += value
    print(coalition_one + ' ' + str(sum_one))
    print(coalition_two + ' ' + str(sum_two))


dictionariy = get_election_results()
print(number_of_votes_are_pass(dictionariy))
dictionariy = calc_mandatim(dictionariy)
coalition_train(dictionariy)

pass: OnlyTovTovwith: 41075
pass: NilsIsALeaderwith: 63702
not pass: DorothyForMidgeLandwith: 23071
pass: MidgeLandIsUswith: 68582
pass: SnufkinForCouncilwith: 35830
not pass: WiliWankawith: 26628
not pass: GulliverIsBestwith: 25928
pass: BestForMidgeLandwith: 40144
not pass: FingeroniBringsPeacewith: 22457
not pass: GreenMidgeLandwith: 5983
249333
OnlyTovTov 20
NilsIsALeader 31
MidgeLandIsUs 33
SnufkinForCouncil 17
BestForMidgeLand 19
OnlyTovTov NilsIsALeader SnufkinForCouncil  68
OnlyTovTov MidgeLandIsUs SnufkinForCouncil  70


<p style="text-align:right;direction:rtl;">
הדפיסו את המילון שקיבלתם. ודאו שסך המנדטים מסתכם ל־120.</p>

<p style="text-align:right;direction:rtl;">
כדי להרכיב קואליציה דרושים 61 מנדטים.<br>
המפלגות MidgeLandIsUs ו־NilsIsALeader לא מוכנות לשבת יחד בקואליציה.<br>
כתבו פונקציה המקבלת את המילון המכיל את תוצאות המנדטים ומדפיסה קואליציה אפשרית עבור NilsIsALeader וקואליציה אפשרית עבור MidgeLandIsUs.
</p> 

## <p style="text-align:right;direction:rtl;">Leveraging Simple Dictionaries</p>

### <p style="text-align:right;direction:rtl;">הקדמה</p>

<p style="text-align:right;direction:rtl;">
    <dfn>אוטומט סופי דטרמיניסטי</dfn> הוא מודל מתמטי שמגדיר אוסף של מצבים, וכללי מעבר ביניהם.<br>
    נפתח בדוגמה:<br>
    <figure>
        <img src="images/even_dfa.svg"  style="max-width:100%; margin-right: auto; margin-left: auto; text-align: center;" width="450px" alt="תמונת דוגמה לאוטומט סופי דטרמיניסטי. בצד ימין ישנו חץ שעליו כתוב 'התחלה'. החץ מצביע לעיגול שבו יש את המילים 'אי־זוגי', וממנו יוצא חץ לכיוון עיגול כפול שבו כתוב 'זוגי'. על החץ כתוב 'הספרה 2, 4, 6, 8 או 0'. מהעיגול הכפול יוצא חץ חזרה לעיגול הראשון, עליו כתוב 'הספרה 1, 3, 5, 7 או 9'. מהעיגול 'זוגי' יוצא חץ לעצמו עם המספרים 2, 4, 6, 8 ו־0. מהעיגול 'אי־זוגי' יוצא חץ לעצמו עם המספרים 1, 3, 5, 7, 9. ">
        <figcaption style="text-align: center; direction: rtl;">אוטומט סופי דטרמיניסטי שבודק האם מספר מסוים הוא מספר זוגי.</figcaption>
    </figure>
</p>

<p style="text-align:right;direction:rtl;">
    לפניכם מכונת מצבים שמקבלת מספר, ובודקת האם הוא זוגי.<br>
    כל מצב מיוצג בתמונה על ידי עיגול, וכללי המעבר ביניהם מיוצגים באמצעות חצים.
</p>

<p style="text-align:right;direction:rtl;">
    המכונה מתחילה מ"מצב התחלתי" נתון כלשהו, במקרה שלנו "מספר אי זוגי".<br>
    כל פעם שמתקבל קלט חדש, המכונה מחליטה לאיזה מצב לעבור לפי כללי המעבר הרשומים על החצים.<br>
    לדוגמה, עבור הקלט 2 ואז 7 ואז 4 (274) מכונת המצבים המופיעה מעלה תבצע את הפעולות הבאות:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>תכנס למצב ההתחלתי "מספר אי זוגי".</li>
    <li>תקרא את הספרה הראשונה במספר – 2, ותחליט לעבור (לפי החץ) למצב "מספר זוגי".</li> 
    <li>תקרא את הספרה השנייה במספר – 7, ותחליט לעבור (לפי החץ) למצב "מספר אי־זוגי".</li> 
    <li>תקרא את הספרה השלישית במספר – 4, ותחליט לעבור (לפי החץ) למצב "מספר זוגי".</li>
</ol>

<p style="text-align:right;direction:rtl;">
    לכל מצב באוטומט סופי דטרמיניסטי יש "סוג מצב": מקבל או לא מקבל.<br>
    מצב מקבל מסומן בעיגול כפול, כמו המצב "מספר זוגי" באוטומט דלעיל.<br>
    אם בסוף הריצה הגענו למצב מקבל, סימן שהקלט תקין, ואם לא – סימן שהקלט אינו תקין.<br>
    באוטומט הדוגמה שלנו כל קלט מספרי זוגי הוא תקין, וכל מספרי אי־זוגי נחשב ללא תקין.
</p>

<p style="text-align:right;direction:rtl;">
    כל אפשרות הכלולה במצבי המעבר נקראת <dfn>אות</dfn>. אוסף כל האותיות הכלולות באוטומט מסוים נקרא <dfn>א"ב הקלט</dfn>.<br>
    שימו לב שבאוטומט יכולה להיות כמות גדולה של מצבים וכללי־מעבר, ובכל אוטומט יכול להיות יותר ממצב מקבל אחד.<br>
    כמו כן, "אות" בא"ב יכולה לכלול יותר מתו אחד.
</p>

### <p style="text-align:right;direction:rtl;">תרגיל</p>

<p style="text-align:right;direction:rtl;">
   כתבו תוכנה שקוראת קובץ ויוצרת ממנו אוטומט סופי דטרמיניסטי.<br>
   בקובץ ישנו כל המידע הדרוש לבניית האוטומט, ורשימת קלטים. דוגמה לקובץ שכזה מופיעה מטה.<br>
   הדפיסו עבור כל קלט האם לאחר הרצתו האוטומט הגיע למצב מקבל או למצב לא מקבל.
</p>

<p style="text-align:right;direction:rtl;">
    הקובץ שהתוכנה שלכם אמורה לקרוא ונועד לתאר את האוטומט המופיע למעלה נראה כך:
</p>

<p style="text-align:right;direction:rtl;">
    על התוכנה שבניתם להדפיס במקרה הזה:
</p>

with open('resources/automat.txt', 'r') as file_automat:
    temp = []
    for line in file_automat.readlines():
        line = line.replace('\n', '')
        if ':' in line:
            line = line.replace(':', '')
        if ',' in line:
            line = line.replace(',', '')
        line = line.split()
        temp += line
print(temp)
diction = {}
dict_one = {}
dict_two = {}
dict_three = {}
dict_four = {}
print(temp)
new = []
i = 0
if temp[i] == 'Definitions':
    if temp[i] == '\n':
        new_val = temp[i]
    i += 1
    while temp[i] != 'Start':
        if temp[i] == 'States':
            val = temp[i]
        else:
            new += [temp[i]]
        i += 1
    dict_one.update({val: new})
    new = []
    while temp[i] != 'Accepting':
        if temp[i] == 'Start':
            val = temp[i]
        else:
            new += [temp[i]]
        i += 1
    dict_two.update({val:new})
    new = []
    while temp[i] != 'Transitions':
        if temp[i] == 'Accepting':
            val = temp[i]
        else:
            new += [temp[i]]
        i += 1
    dict_three.update({val: new})
diction.update({new_val: [dict_one, dict_two, dict_three]})
print(diction)
print(dict_one)
print(dict_two)
print(dict_three)
new = []
if temp[i] == 'Transitions':
    val = temp[i]
    i += 1
    while temp[i] != 'Inputs':
        new += [temp[i]]
        i += 1 
dict_four.update({val: new})
print(dict_four)
with open('resources/automat.txt', 'r') as file_automat:
    temp = file_automat.readlines()
i=0
length = len(temp)
while temp[i] != 'Inputs:\n' :
    print(i)
    i+=1
    length -= 1
if temp[i]== 'Inputs:\n':
    temp[i] = temp[i].replace(':', '').replace('\n', '')
    val = temp[i]
i += 1
length -= 1
print(length)
print(val)
new = []
dict_five = {}
while length:
    temp[i] = temp[i].replace(',', '').split()
    temp[i] += '\n'
    new += temp[i]
    i += 1
    length -= 1
dict_five.update({val: new})
print(dict_five)
state = dict_two['Start']
accept = dict_three['Accepting']
temp = dict_five['Inputs']
trans = dict_four['Transitions']
leng = len(temp)
length = len(trans)
i=0
j=0
flag_find = 0
# while leng:
while length:
    if state[0] in trans[j]:
        while trans[j] != '->':
            if temp[i] in trans[j]:
                flag_find = 1
            j += 1
       
        if flag_find == 1:
            state[0] = trans[j+1]
        j = 0
    j+=1
    if temp[i] == '\n':
        if state[0] in accept[0]:
            print('True')
        else:
            print('False')
    i += 1
    length -= 1
#     leng -= 1

### <p style="text-align:right;direction:rtl;">הסבר עבור הדוגמה</p>

In [36]:
with open('resources/automat.txt', 'r') as file_automat:
    temp ={}
    for line in file_automat.readlines():
        if 'Definition' in line:
            temp['Definition'] = 

yes


<p style="text-align:right;direction:rtl;">
    הפסקה הראשונה תמיד תראה כך:<br>
    הקובץ תמיד יפתח במילה <q dir="ltr">Definitions:</q>.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">States:</q>, ואחריה כל המצבים הקיימים במכונת המצבים, מופרדים בפסיק.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">Start:</q>, ואחריה המצב ההתחלתי של המכונה.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">Accepting:</q>, ואחריה כל המצבים שמוגדרים כמצבים מקבלים במכונת המצבים, מופרדים בפסיק.<br>
</p>

<p style="text-align:right;direction:rtl;">
    הפסקה השנייה תמיד תראה כך:<br>
    הפסקה תמיד תפתח במילה <q dir="ltr">Transitions:</q>, ואחריה רשימה של כללי מעבר.<br>
    כל כלל מעבר מורכב מהמצב ממנו יוצאים, המצב אליו מגיעים ומה הקלט ש"מפעיל" את המעבר.<br>
    השורה תכתב באופן הבא, כאשר X הוא המצב ממנו יוצאים, Z הוא המצב אליו מגיעים ו־Y היא רשימת הקלטים שמפעילה את מצב המעבר:<br>
    X: Y -> Z<br>
</p>

<p style="text-align:right;direction:rtl;">
    הפסקה השלישית תמיד תראה כך:<br>
    הפסקה תמיד תפתח במילה <q dir="ltr">Inputs:</q>, ואחריה רשימה של קלטים <strong>תקינים</strong>.<br>
    כל שורת קלט תופיע בשורה נפרדת בקובץ, ותכלול את רצף הקלטים שיועברו לאוטומט אחד אחרי השני, כאשר הם מופרדים בפסיק.
</p>

<p style="text-align:right;direction:rtl;">
    עבור הקלט הראשון, האוטומט יתחיל מהמצב ההתחלתי q0. האות 1 תשאיר אותו שם, כך גם 3, כך גם 1. מצב לא מקבל.<br>
    עבור הקלט השני, האוטומט יתחיל מהמצב ההתחלתי q0. האות 2 תעביר אותו ל־q1, האות 7 תעביר אותו ל־q0, האות 4 תעביר אותו חזרה ל־q1. מצב מקבל.<br>
    עבור הקלט השלישי, האוטומט יתחיל מהמצב ההתחלתי q0. האות 2 תעביר אותו ל־q1, האות 2 תשאיר אותו ב־q1 בכל המופעים הבאים שלה. מצב מקבל.<br>
</p>

### <p style="text-align:right;direction:rtl;">דוגמה נוספת</p>

<p style="text-align:right;direction:rtl;">
    זוהי דוגמה לקובץ שמתאר את האוטומט המופיע <a href="https://en.wikipedia.org/wiki/Deterministic_finite_automaton#/media/File:DFA_example_multiplies_of_3.svg">כאן</a>.
</p>